server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -330,20 +381,45 @@ public ClusterMetadataManifest writeIncrementalMetadata(
             }
             previousStateIndexMetadataVersionByName.remove(indexMetadata.getIndex().getName());
         }
+        UploadedMetadataResults uploadedMetadataResults;
+        boolean firstUpload = !previousManifest.hasMetadataAttributesFiles();
+        // For migration case from codec V0 or V1 to V2, we have added null check on metadata attribute files,
+        // If file is empty and codec is 1 then write global metadata.
+        if (firstUpload) {
+            uploadedMetadataResults = writeMetadataInParallel(clusterState, toUpload, clusterState.metadata().customs(), true, true, true);
+        } else {
+            uploadedMetadataResults = writeMetadataInParallel(
+                clusterState,
+                toUpload,
+                customsToUpload,
+                updateCoordinationMetadata,
+                updateSettingsMetadata,
+                updateTemplatesMetadata
+            );
+        }
 
-        List<UploadedIndexMetadata> uploadedIndexMetadataList = writeIndexMetadataParallel(clusterState, toUpload);
-        uploadedIndexMetadataList.forEach(
+        uploadedMetadataResults.uploadedIndexMetadata.forEach(
             uploadedIndexMetadata -> allUploadedIndexMetadata.put(uploadedIndexMetadata.getIndexName(), uploadedIndexMetadata)
         );
-
+        allUploadedCustomMap.putAll(uploadedMetadataResults.uploadedCustomMetadataMap);
+        previousStateCustomMap.keySet().forEach(allUploadedCustomMap::remove);
```

Comment:
Can you add a test where we verify the addition, deletion, updation of custom metadata ? The new manifest file should be checked to verify the same.

Reply:
Ack, will add.